api/src/main/java/bio/terra/tanagra/common/Paginator.java

```diff
@@ -0,0 +1,82 @@
+package bio.terra.tanagra.common;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.tanagra.proto.utils.pagination.PageSizeOffset;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.InvalidProtocolBufferException;
+import java.util.Base64;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Implements simple pagination for sorted lists in memory. */
+public final class Paginator<T> {
+  private Paginator() {}
+
+  /**
+   * Returns a {@link Page} for the results.
+   *
+   * @param allResults the sorted list of all results being paginated.
+   * @param pageSize the maximum number of results to have per page.
+   * @param pageToken the token of the page to retrieve. If null or empty, the first page is
+   *     retrieved.
+   */
+  public static <T> Page<T> getPage(List<T> allResults, int pageSize, @Nullable String pageToken) {
+    Preconditions.checkArgument(pageSize > 0, "pageSize must be greater than 0 to paginate.");
+    if (pageToken == null || pageToken.isEmpty()) {
+      return getPageInternal(
+          allResults, PageSizeOffset.newBuilder().setOffest(0).setPageSize(pageSize).build());
+    }
+    PageSizeOffset pageSizeOffset = decode(pageToken);
+    if (pageSizeOffset.getPageSize() != pageSize) {
+      throw new BadRequestException(
+          String.format(
+              "Token page size '%d' does not match input page size '%d'.",
+              pageSizeOffset.getPageSize(), pageSize));
+    }
+    return getPageInternal(allResults, decode(pageToken));
+  }
+
+  public static <T> Page<T> getPageInternal(List<T> allResults, PageSizeOffset pageSizeOffset) {
+    Preconditions.checkArgument(
+        pageSizeOffset.getOffest() < allResults.size() || pageSizeOffset.getOffest() == 0,
+        "page offset out of bounds.");
+    int offset = pageSizeOffset.getOffest();
+    int pageLimit = pageSizeOffset.getOffest() + pageSizeOffset.getPageSize();
+    if (pageLimit >= allResults.size()) {
+      List<T> results = allResults.subList(offset, allResults.size());
+      // This page includes the last of the results. Use an empty string for the next page token.
+      return Page.create(results, "");
+    }
+    PageSizeOffset nextPageSizeOffset = pageSizeOffset.toBuilder().setOffest(pageLimit).build();
+    List<T> results = allResults.subList(offset, pageLimit);
+    return Page.create(results, encode(nextPageSizeOffset));
+  }
+
+  private static String encode(PageSizeOffset pageSizeOffset) {
+    return Base64.getEncoder().encodeToString(pageSizeOffset.toByteArray());
+  }
+
+  private static PageSizeOffset decode(String pageToken) {
```

Comment:
FYI: not relevant yet, since you don't really have any parameterization on your list endpoint, but one pattern I've used in the past:

- take a short hash of the request (sans pageSize, pageToken)
- add this to the page token payload
- validate this against incoming requests that include a token

This allows you to reject requests that break the pagination contract - i.e. attempt to intentionally or accidentally change filter parameters (e.g. for listEntities, say they change the underlayName) in the middle of a pagination stream.

Reply:
Thanks, added a field to the proto and updated the Paginator.

Reply:
Nice!